net: lib: http_server: add static file system resource

MeisterBob opened the pull request:
adds file system as a resource for the http_server which serves static (gzipped) files from a file system to the client.

jukkar commented on include/zephyr/net/http/server.h:
```diff
@@ -115,6 +118,16 @@ struct http_resource_detail_static {
 BUILD_ASSERT(offsetof(struct http_resource_detail_static, common) == 0);
 /** @endcond */
 
+struct http_resource_detail_static_fs {
+	struct http_resource_detail common;
+	const char *fs_path;
```
Please add doxygen comments to these.

jukkar commented on subsys/net/lib/http/http_server_http1.c:
```diff
@@ -256,6 +257,128 @@ static int dynamic_post_req(struct http_resource_detail_dynamic *dynamic_detail,
+		return ret;
+	}
+
+	/* get filename amd content-type from url */
```
typo s/amd/and/

jukkar commented on subsys/net/lib/http/http_server_http1.c:
```diff
@@ -256,6 +257,128 @@ static int dynamic_post_req(struct http_resource_detail_dynamic *dynamic_detail,
+	/* get filename amd content-type from url */
+	len = strlen(client->url_buffer);
+	if (len > 1) {
+		if (strcmp(&client->url_buffer[len - 5], ".html") == 0) {
+			content_type = "text/html";
+		} else if (strcmp(&client->url_buffer[len - 3], ".js") == 0) {
+			content_type = "text/javascript";
+		} else if (strcmp(&client->url_buffer[len - 4], ".css") == 0) {
+			content_type = "text/css";
+		} else if (strcmp(&client->url_buffer[len - 4], ".jpg") == 0) {
+			content_type = "image/jpeg";
+		} else if (strcmp(&client->url_buffer[len - 4], ".png") == 0) {
+			content_type = "image/png";
+		} else if (strcmp(&client->url_buffer[len - 4], ".svg") == 0) {
+			content_type = "image/svg+xml";
```
Instead of this, we could have an static const array of extensions and content_types, and we could just go through them in a helper function which would be called here. Then adding new content types would be just add new entries to the said array.

MeisterBob replied:
I have placed the content-types in an iterable section so that users can add new content types themselves.

jukkar commented on subsys/net/lib/http/http_server_http1.c:
```diff
@@ -256,6 +257,128 @@ static int dynamic_post_req(struct http_resource_detail_dynamic *dynamic_detail,
+	int ret;
+	struct fs_dirent dirent;
+	struct fs_file_t file;
+	char fname[256];
```
Do we really need to have this long filename here, it causes lot of pressure to the used stack size.

MeisterBob replied:
`fname` needs to consume `fs_path` and and the `url_buffer`. The `url_buffer` is 256 bytes long, to be safe we would need 256 + `strlen(fs_path)`. I changed the size to `HTTP_SERVER_MAX_URL_LENGTH`, so the user can reduce the buffer size when he runs short on stack size.

jukkar reviewed (commented):
I think we also need to support HTTP2 here.
Also tests for this are needed.

rlubos commented:
I agree HTTP2 support should be added too.

MeisterBob commented:
Thanks for your feedback. I addressed the issues you had and will add HTTP2 next.

jukkar commented on subsys/net/lib/http/http_server_http1.c:
```diff
@@ -256,6 +257,135 @@ static int dynamic_post_req(struct http_resource_detail_dynamic *dynamic_detail,
+HTTP_SERVER_CONTENT_TYPE(png, "image/png")
+HTTP_SERVER_CONTENT_TYPE(svg, "image/svg+xml")
+
+static void _get_content_type_from_extension(char *url, size_t url_len, char *content_type,
```
BTW, no need to prefix static functions with `_`, I also recall that the single underscore is meant to be used in C-library only. Anyway, there is probably some tool that will start to complain about these at some point so better to remove the underscore from function name.

MeisterBob replied:
As it is needed for http2 as well I renamed it to `http_server_get_content_type_from_extension` and moved it to `http_server_core.c`.

jukkar commented on subsys/net/lib/http/iterables.ld:
```diff
@@ -0,0 +1 @@
+ITERABLE_SECTION_ROM(http_content_type, 4)
```
Please add newline this this file.

jukkar replied:
Please add copyright + spdx info to this file.
Also related to change in commit dacb3dbfeb5d34ffc02c2e2f547dbd569ace008e,
please add this before the macro call.
```
#include <zephyr/linker/iterable_sections.h>
```

MeisterBob commented:
added a view tests

jukkar commented on subsys/net/lib/http/http_server_core.c:
```diff
@@ -704,6 +705,42 @@ struct http_resource_detail *get_resource_detail(const char *path,
+	{
+		char *ext = &url[url_len - ct->extension_len];
+
+		if (0 == strncmp(ext, ct->extension, ct->extension_len)) {
```
Please don't use this style, we always do the check other way around, so `strncmp() == 0`

jukkar commented on subsys/net/lib/http/http_server_http1.c:
```diff
@@ -256,6 +257,116 @@ static int dynamic_post_req(struct http_resource_detail_dynamic *dynamic_detail,
+	ret = http_server_find_file(fname, sizeof(fname), &file_size, &gzipped);
+	if (ret < 0) {
+		LOG_ERR("fs_stat %s: %d", fname, ret);
+		static const char not_found_response[] = "HTTP/1.1 404 Not Found\r\n"
```
Please put the log_err after this line

jukkar commented on subsys/net/lib/http/http_server_http2.c:
```diff
@@ -371,6 +371,108 @@ static int handle_http2_static_resource(
+			LOG_DBG("Cannot write to socket (%d)", ret);
+			goto out;
+		}
+		if (ret < 0) {
```
This check seems to a duplicate

jukkar commented on tests/net/lib/http_server/common/src/main.c:
```diff
@@ -12,30 +12,37 @@
 	{
-			.type = HTTP_RESOURCE_TYPE_STATIC,
-			.bitmask_of_supported_http_methods = BIT(HTTP_GET),
+		.type = HTTP_RESOURCE_TYPE_STATIC,
```
Please place the tests to a separate commit

rlubos commented on subsys/net/lib/http/http_server_http1.c:
```diff
@@ -256,6 +257,116 @@ static int dynamic_post_req(struct http_resource_detail_dynamic *dynamic_detail,
+
+	/* get filename and content-type from url */
+	len = strlen(client->url_buffer);
+	if (len > 1) {
```
I'm not sure I follow, why special treatment of `len == 1`?

MeisterBob replied:
`len == 1` means, that the requested path is just the leading `/` and we can just use `index.html` as filename. I added a comment to make that clear.

rlubos commented on subsys/net/lib/http/http_server_http2.c:
```diff
@@ -869,6 +971,12 @@ int handle_http1_to_http2_upgrade(struct http_client_ctx *client)
 				goto error;
 			}
+		} else if (detail->type == HTTP_RESOURCE_TYPE_STATIC_FS) {
+			ret = handle_http2_static_fs_resource(
```
It should also be supported from `handle_http_frame_headers()`, this handles only HTTP upgrade case.

rlubos commented on subsys/net/lib/http/http_server_http2.c:
```diff
@@ -371,6 +371,108 @@ static int handle_http2_static_resource(
+	while (remaining > 0) {
+		len = fs_read(&file, tmp, sizeof(tmp));
+
+		ret = send_data_frame(client, tmp, len, frame->stream_identifier, 0);
```
This could be optimized a bit - do `remaining -= len` first, and if `remaining == 0` set `HTTP2_FLAG_END_STREAM` in the last data frame. That way there will be no need to send an empty frame afterwards.

rlubos reviewed (commented):
Not asking to do in this PR, but with this model it should also be possible to support POST/DELETE with filesystem, correct?

MeisterBob commented on subsys/net/lib/http/http_server_core.c:
```diff
@@ -704,6 +705,42 @@ struct http_resource_detail *get_resource_detail(const char *path,
 	return NULL;
 }
 
+int http_server_find_file(char *fname, size_t fname_len, size_t *file_size, bool *gzipped)
```
shall I use `fname_size` instead?

rlubos replied:
Sounds better

mt-gjoerges committed:
net: lib: http_server: add static fs resource

adds filesystem as a resource for the http_server which serves static
(gzipped) files from a filesystem to the client.

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

mt-gjoerges committed:
tests: net: http_server: Add static fs tests

Add a test that checks whether the extensions are correctly assigned to
content types.

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

MeisterBob commented on subsys/net/lib/http/http_server_http1.c:
```diff
@@ -256,6 +257,116 @@ static int dynamic_post_req(struct http_resource_detail_dynamic *dynamic_detail,
+	/* read and send file */
+	remaining = file_size;
+	while (remaining > 0) {
+		len = fs_read(&file, tmp, sizeof(tmp));
```
`dynamic_get_req()` and `dynamic_post_req()` also use an `http_response` and `tmp` buffer, whereby `http_response` is hidden behind the `SEND_RESPONSE` macro.

rlubos reviewed (dismissed):
Please fix the CI failure.

jukkar commented on subsys/net/lib/http/http_server_core.c:
```diff
@@ -704,6 +712,44 @@ struct http_resource_detail *get_resource_detail(const char *path,
+	size_t url_len = strlen(url);
+
+	HTTP_SERVER_CONTENT_TYPE_FOREACH(ct)
+	{
```
Please place the starting brace to previous line

MeisterBob replied:
do you want me to add the `HTTP_SERVER_CONTENT_TYPE_FOREACH` to the `ForEachMacros` list in zephyr's .clang-format file?

jukkar replied:
Is following zephyr style causing some issue with clang-format?

MeisterBob replied:
clang-format isn't exactly formatting wrong, it just doesn't know about FOREACH macros and seems to think it's a function and therefore moves the opening brace to a newline.

jukkar reviewed (commented):
Looks good, couple of minor nits from me.
